Add checksum aggregate Presto function

kgpai opened the pull request:
- Also fix a bug in PrestoHasher for map type.
- Also add BaseVector::clearAllNulls method.

kagamiori commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,162 @@
+namespace facebook::velox::aggregate {
+template <TypeKind kind>
+int64_t getHash(const DecodedVector& vector, const vector_size_t row) {
+  return getHash<kind>(vector, row);
```
Does `getHash` call itself in this general case?

kagamiori reviewed (commented):
Hi kgpai, Thank you for adding this function! I'm trying to understand the code and have a few general questions:
* Which checksum algorithm do we use? Is there any special properties about it such as the range of the checksum value?
* I just started looking at aggregation functions and noticed that some functions inherit `SimpleNumericAggregate` while some others inherit `exec::Aggregate`. What's the difference between them?

kagamiori commented on velox/functions/prestosql/aggregates/PrestoHasher.cpp:
```diff
@@ -0,0 +1,162 @@
+namespace facebook::velox::aggregate {
+template <TypeKind kind>
+int64_t getHash(const DecodedVector& vector, const vector_size_t row) {
+  return getHash<kind>(vector, row);
```
Would `getHash` call itself in this general case?

kgpai replied:
It should switch to the instantiated version, but this code isnt required. let me remove it.

kagamiori commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+
+      if (!isNull) {
+        rows.template applyToSelected(
+            [&](vector_size_t row) { computeHash(groups[row], decoded, 0); });
```
Is `decoded` still accessed through `valueAt<T>(row)` at each row? It could be better if we can assign `valueAt<T>(0)` to a local variable before the `applyToSelected` loop and use this variable directly in the loop.

kgpai replied:
Not sure i understood, but the value is accessed via getHash function , which takes decoded vector and then valueAt. The reason I am not assigning to a local variable is because that doesnt work for complex, nested types - as cost of materializing/copy can be high.

mbasmanova replied:
@kgpai I think @kagamiori has a point. With the existing code, computeHash ends up calling getHash(decoded, 0) again an again for each row. Wouldn't it be better to calculate that value once, then add it to existing hashes in a loop.

kagamiori commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+        getHash, vector.base()->typeKind(), vector, row);
+    auto result = *exec::Aggregate::value<int64_t>(group);
+
+    result += times * hashVal * kPrime64;
```
Might it overflow during the calculation? Or do we know it won't happen?

kgpai replied:
Shouldnt matter, it acts as a ring.

kgpai commented:
I have moved hash logic to this PR : #804 . After that is approved will merge those changes here.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    for (auto i : indices) {
+      *this->template value<int64_t>(groups[i]) = (int64_t)0;
```
*exec::Aggregate::value<int64_t>(groups[i]) = 0

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto* vector = (*result)->as<FlatVector<int64_t>>();
```
If result type is VARBINARY as specified in the constructor, then result vector will be FlatVector<StringView>, not FlatVector<int64_t>.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+        clearNull(rawNulls, i);
+        int64_t val = *value<int64_t>(group);
+        auto buffer =
+            AlignedBuffer::allocate<char>(sizeof(int64_t), vector->pool());
```
a reinterpret_cast should be sufficient

mbasmanova reviewed (commented):
@kgpai Some quick comments.

kgpai commented:
* Added PrestoHasher for computing hashes for any Vector type 
* Added test cases for different vector types. 
* Addressed some PR comments.

kgpai commented:
* Added PrestoHasher for computing hashes for any Vector type 
* Added test cases for different vector types. 
* Addressed some PR comments.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+      bool /**/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    BufferPtr hashes =
```
It would be nice to reuse this buffer.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+
+  FOLLY_ALWAYS_INLINE PrestoHasher* getPrestoHasher(TypePtr typePtr) {
+    if (prestoHasher_ == nullptr) {
+      prestoHasher_ = std::make_unique<PrestoHasher>(typePtr);
```
std::move(typePtr)

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+
+ private:
+  FOLLY_ALWAYS_INLINE void
+  computeHash(char* group, const int64_t* hashes, const vector_size_t& row) {
```
Looks like this function could take a group and a single hash: `(char* group, int64_t hash)`

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,225 @@
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected([&](vector_size_t i) {
+      *value<int64_t>(groups[i]) = *(getHashValue(rawValues[i]));
```
I'm wondering if this is correct. It looks like the accumulator for each group is overwritten instead of the existing value being combined with the incoming value.

mbasmanova reviewed (commented):
Some comments.

kgpai commented:
Hi Wei, Apologies for the late answer to your questions:

> 1. Which checksum algorithm do we use? Is there any special properties about it such as the range of the checksum value?

We use XXHash's 64 bit implementation. You can read about it here : https://cyan4973.github.io/xxHash/ . It covers the whole 64 bit range space. 

> I just started looking at aggregation functions and noticed that some functions inherit SimpleNumericAggregate while some others inherit exec::Aggregate. What's the difference between them?

exec::Aggregate is the main base class while SimpleNumericAggregate is a specialized instance of it. SimpleNumericAggregate can be used when you can save your accumulators using simple primitive types and presumes that the intermediate and final aggregates are same type. That is not the case for checksum aggregate since we need to return Varbinary.

kagamiori commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,231 @@
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+    int64_t result = 0;
+    if (vector->mayHaveNulls()) {
```
I'm just curious about the difference between `addIntermediateResults()` and `addSingleGroupIntermediateResults()`. I thought the two functions were similar except the former deals with multiple groups whereas the latter deals with a single group. So I thought they either both check for null or both do not. Are they actually more different?

kgpai replied:
Thats a good question/observation Wei, I have noticed that singe group intermediate results can have nulls in them : for e.g if you have multiple splits, some filter condition , then no aggregation takes place in some of the splits, there single group intermediate results is called once / split (and can be called potentially on a different machine). However I havent noticed this for addIntermediateResults. countAggregate also follows the same pattern. Maybe I should add null checks in addIntermediateResults too.

kagamiori reviewed (commented):
My comments were addressed in Phabricator, so it looks good to me. But I would suggest you wait for an accept from someone else because I'm not yet an expert in aggregate function implementations.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+
+    for (int32_t i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      if (isNull(group)) {
```
checksum is never null, right? remove this branch

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+      if (isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        vector->set(i, *value<int64_t>(group));
```
consistency: similar logic above uses rawValues; use the same here

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
```
make sure to clean any nulls set in result (in case of reuse)

```
vector->clearNulls(0, vector->size());
```

It might be helpful to add BaseVector::clearAllNulls() method.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
```
/**/ -> /* mayPushDown */

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
```
Why do we need this check? PrestoHasher takes care of nulls, no? Can't we just use rawHashes for all rows?

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
```
consistency: this code uses asUnchecked, but other places use as. Use asUnchecked in all places as it is more efficient. Also, other places add VELOX_CHECK(vector), but this one doesn't. Make it consistent.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected(
+        [&](vector_size_t i) { *value<int64_t>(groups[i]) += rawValues[i]; });
```
i -> row

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
```
/**/ -> /* mayPushDown */

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
```
same question as above

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
```
same as above

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    clearNull(group);
```
Is this needed? checksum cannot be null, can it?

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+      bool /**/) override {
+    clearNull(group);
+
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
```
consistency: other places use temp variable for args[0] and check that vector is not null

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+    int64_t result = 0;
+    if (vector->mayHaveNulls()) {
```
consistency: other places do not have a separate path for no-nulls input

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+  }
+
+  FOLLY_ALWAYS_INLINE BufferPtr& getHashBuffer(
+      vector_size_t size,
```
The callers seem to be passing size as the number of entries, but hashes_->size() is the number of bytes, e.g. 8 * number-of-entries. Hence, the check below seems wrong.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,228 @@
+          const TypePtr& /*resultType*/) -> std::unique_ptr<exec::Aggregate> {
+        VELOX_CHECK_EQ(argTypes.size(), 1, "{} takes one argument", name);
+
+        if (step == core::AggregationNode::Step::kFinal or
```
nit: consider using isPartialOutput(step) for brevity

mbasmanova commented on velox/functions/prestosql/aggregates/tests/ChecksumAggregateTest.cpp:
```diff
@@ -0,0 +1,253 @@
+ * limitations under the License.
+ */
+#include <boost/algorithm/string/join.hpp>
+#include "velox/exec/AggregationHook.h"
```
is this include used?

mbasmanova reviewed (commented):
@kgpai Looks good overall. I have a few questions about null handling.

kgpai commented:
- Added clearAllNulls() to BaseVector
- Fixes as per comments (Thanks for pointing out the hash->size() bug @mbasmanova !)
- Modified createMapOfArraysVector to be generic for any kind of map.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+
+namespace {
+
+/// ChecksumAggregate computes an order-insensitive checksum of the input
```
nit: No need to repeat "ChecksumAggregate" as it is obvious. Just say "Computes ..."

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    for (auto i : indices) {
```
This function needs to call `setAllNulls(groups, indices);` to initialize null flags for the new groups.

In GROUP BY case, new groups are added only when there is a row with matching grouping keys, but in case of a global aggregation the only group is added even if there is no data. Hence, in case of global aggregation, the group may never see any data.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    vector->clearAllNulls();
```
In case of global aggregation, the group may be null, hence, we cannot assume that all groups have non-null values. It would be nice to add a test for global aggregation over empty dataset.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto vector = (*result)->asUnchecked<FlatVector<int64_t>>();
```
VELOX_CHECK(vector); is missing

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    vector->clearAllNulls();
```
same as above

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+    const auto& arg = args[0];
+    auto vector = arg->asUnchecked<FlatVector<int64_t>>();
+    VELOX_CHECK(vector);
+    auto rawValues = vector->rawValues();
```
There can be nulls in case of global aggregation over empty dataset, no?

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+    clearNull(group);
+    auto result = *value<int64_t>(group);
+    result += hash * PRIME64_1;
+    *value<int64_t>(group) = result;
```
nit: `*value<int64_t>(group) += hash * PRIME64_1;`

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+
+ private:
+  FOLLY_ALWAYS_INLINE void computeHash(char* group, const int64_t hash) {
+    clearNull(group);
```
this call can be moved to the caller as it is shared between computeHash and computeHashForNull

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,215 @@
+      velox::memory::MemoryPool* pool) {
+    // hashes_->size() is in bytes.
+    if (hashes_ == nullptr || hashes_->size() < (size * sizeof(int64_t))) {
+      hashes_ = AlignedBuffer::allocate<int64_t>(size, pool);
```
use AlignedBuffer::reallocate if hashes_ is not null; in case there is sufficient capacity in the buffer, reallocate will reuse memory instead of allocating new one.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/ChecksumAggregateTest.cpp:
```diff
@@ -0,0 +1,254 @@
+                     .planNode();
+      assertQuery(
+          agg,
+          fmt::format("VALUES (CAST(\'{}\' AS VARBINARY))", expectedChecksum));
```
to_base64 returns a varchar; why do we cast expected checksum to varbinary here?

mbasmanova reviewed (commented):
@kgpai Looks good to me. Does this work with values of type TIMESTAMP WITH TIME ZONE ? CC: @kagamiori

kgpai commented:
* Fixed bug for global aggregation over empty dataset.
* Added test case for global aggregation over empty dataset.
* Added test case for TimestampwithTimezone type. 
* End to End test with global aggregation over empty dataset returns equivalent results to presto.

kgpai commented:
- Creating a separate PR to add support for TimestampWithTimezone in PrestoHasher. Will link that here when created.

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,235 @@
+      if (isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        vector->setNull(i, false);
```
In most cases, there will be no nulls. Hence, it would be more efficient to clear all nulls before the loop and call setNull only if isNull(group).

```
vector->clearAllNulls();
for (...) {

   if (isNull(group)) {
      vector->setNull(i, true);
   } else {
       rawValues[i] =...
   }
}
```

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,235 @@
+      auto group = groups[i];
+      if (isNull(group)) {
+        vector->setNull(i, true);
+        continue;
```
nit: use 'else' for consistency with extractValues method

```
   if (isNull(group)) {
      vector->setNull(i, true);
   } else {
       rawValues[i] =...
   }
```

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,235 @@
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      clearNull(groups[row]);
```
nit: use temp variable for groups[row] to avoid copy-paste and make this method consistent with the rest

mbasmanova commented on velox/functions/prestosql/aggregates/ChecksumAggregate.cpp:
```diff
@@ -0,0 +1,235 @@
+    auto vector = arg->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected([&](vector_size_t row) {
```
Perhaps, add a comment explaining why null check is missing here. Also, since these functions can be used outside of Presto, e.g. by XStream, it might be safer to not assume that there will be no nulls in the intermediate results and add a check.

kgpai replied:
In the case of checksum,  I dont see how it should be  possible for intermediate groups to have null value . However I have added a null check.

mbasmanova commented on velox/functions/prestosql/aggregates/tests/ChecksumAggregateTest.cpp:
```diff
@@ -0,0 +1,286 @@
+      5,
+      std::vector<VectorPtr>{timestamp, timezone});
+
+  assertChecksum(timestampWithTzVector, "411aYzf4Rog=");
```
You mentioned that checksum calculation for Timestamp with time zone is wrong. If this tests is not failing, then it means the expected value of the checksum here is wrong. Am I missing something?

mbasmanova commented on velox/vector/tests/VectorTestBase.h:
```diff
@@ -15,6 +15,7 @@
  */
 #pragma once
 
+#include "iostream"
```
this include is not needed, right? let's remove

mbasmanova commented on velox/vector/tests/VectorTestBase.h:
```diff
@@ -451,6 +452,63 @@ class VectorTestBase {
+
+    return std::make_shared<MapVector>(
+        pool_.get(),
+        MAP(BIGINT(), ARRAY(BIGINT())),
```
This type doesn't look right. Can you try this function with T = double or something other than int64_t?

kgpai replied:
Fixed.

mbasmanova commented on velox/vector/tests/VectorTestBase.h:
```diff
@@ -451,6 +452,63 @@ class VectorTestBase {
+  }
+
+  template <typename T>
+  static std::optional<std::vector<std::optional<T>>> makeOptional(
```
What't the motivation for introducing this method ? It seems identical to std::make_optional.

kgpai commented on velox/vector/tests/VectorTestBase.h:
```diff
@@ -451,6 +451,63 @@ class VectorTestBase {
     return velox::test::VectorMaker::flatten(vector);
   }
 
+  // Convenience function to create a vector of type Map[K, ARRAY[K]].
```
No reason, just ease of use I suppose, I will make it different types .

kgpai commented on velox/vector/tests/VectorTestBase.h:
```diff
@@ -451,6 +451,63 @@ class VectorTestBase {
+
+    auto mapValues = makeVectorWithNullArrays(values);
+    auto mapKeys = makeNullableFlatVector<T>(keys);
+    auto size = mapSizes.size();
```
Hi Masha, values stores all the values from all the maps, so values.size() will not be same as mapSizes . values[i].size will be size of individual vector which also wont be right i think.

mbasmanova replied:
Is the following not right?

```
auto size = maps.size();
```

mbasmanova replied:
In other words, here is what I think can also work:

```
auto size = maps.size();

auto offsets = allocateOffsets(size, pool_.get());
auto sizes = allocateSizes(size, pool_.get());

auto rawOffsets = offsets->template asMutable<vector_size_t>();
auto rawSizes = sizes->template asMutable<vector_size_t>();

for (vector_size_t i = 0; i < size; i++) {
      rawSizes[i] = maps[i].size();
      rawOffsets[i] = offset;
      offset += rawSizes[i];
    }
```

mbasmanova replied:
BTW, using allocateOffsets and allocateSizes is safer than using AlignedBuffer::allocate directly.
